src/webgl/p5.Matrix.js

```diff
@@ -1,717 +1,1324 @@
+  var x, y, z;
+
+  if (axis instanceof p5.Vector) {
+    x = axis.x;
+    y = axis.y;
+    z = axis.z;
+  } else if (axis instanceof Array) {
+    x = axis[0];
+    y = axis[1];
+    z = axis[2];
+  }
+
+  // normalize axis
+  var len = x * x + y * y + z * z;
+  if (len === 0.0) {
+    return null;
+  } else if (len !== 1.0) {
+    len = 1.0 / Math.sqrt(len);
+    x *= len;
+    y *= len;
+    z *= len;
+  }
+
+  // rotation coeffs
+  angle = this.assertAngleMode(angle);
+  var sA = Math.sin(angle);
+  var cA = Math.cos(angle);
+  var tA = 1 - cA;
+
+  // rotation matrix
+  var x0 = x * x * tA + cA;
+  var x1 = x * y * tA - z * sA;
+  var x2 = x * z * tA + y * sA;
+
+  var y0 = y * x * tA + z * sA;
+  var y1 = y * y * tA + cA;
+  var y2 = y * z * tA - x * sA;
+
+  var z0 = z * x * tA - y * sA;
+  var z1 = z * y * tA + x * sA;
+  var z2 = z * z * tA + cA;
+
+  // this = this * mRotation
+  if (this.mat4) {
+    // prettier-ignore
+    this.mult([x0, y0, z0, 0, 
+               x1, y1, z1, 0, 
+               x2, y2, z2, 0, 
+               0, 0, 0, 1]);
   } else {
-    //convert mat4 -> mat3
-    this.mat3[0] = matrix.mat4[0];
-    this.mat3[1] = matrix.mat4[1];
-    this.mat3[2] = matrix.mat4[2];
-    this.mat3[3] = matrix.mat4[4];
-    this.mat3[4] = matrix.mat4[5];
-    this.mat3[5] = matrix.mat4[6];
-    this.mat3[6] = matrix.mat4[8];
-    this.mat3[7] = matrix.mat4[9];
-    this.mat3[8] = matrix.mat4[10];
+    // prettier-ignore
+    this.mult([x0, y0, z0, 
+               x1, y1, z1, 
+               x2, y2, z2]);
   }
 
-  this.invert3x3().transpose3x3(this.mat3);
   return this;
 };
 
 /**
- * inspired by Toji's mat4 determinant
- * @method determinant
- * @return {Number} Determinant of our 4x4 matrix
+ * Rotates the matrix around the X-Axis.
+ *
+ * @method rotateX
+ * @param {Number} angle
+ * @chainable
  */
-p5.Matrix.prototype.determinant = function() {
-  var d00 = this.mat4[0] * this.mat4[5] - this.mat4[1] * this.mat4[4],
-    d01 = this.mat4[0] * this.mat4[6] - this.mat4[2] * this.mat4[4],
-    d02 = this.mat4[0] * this.mat4[7] - this.mat4[3] * this.mat4[4],
-    d03 = this.mat4[1] * this.mat4[6] - this.mat4[2] * this.mat4[5],
-    d04 = this.mat4[1] * this.mat4[7] - this.mat4[3] * this.mat4[5],
-    d05 = this.mat4[2] * this.mat4[7] - this.mat4[3] * this.mat4[6],
-    d06 = this.mat4[8] * this.mat4[13] - this.mat4[9] * this.mat4[12],
-    d07 = this.mat4[8] * this.mat4[14] - this.mat4[10] * this.mat4[12],
-    d08 = this.mat4[8] * this.mat4[15] - this.mat4[11] * this.mat4[12],
-    d09 = this.mat4[9] * this.mat4[14] - this.mat4[10] * this.mat4[13],
-    d10 = this.mat4[9] * this.mat4[15] - this.mat4[11] * this.mat4[13],
-    d11 = this.mat4[10] * this.mat4[15] - this.mat4[11] * this.mat4[14];
+p5.Matrix.prototype.rotateX = function(angle) {
+  angle = this.assertAngleMode(angle);
+  var c = Math.cos(angle);
+  var s = Math.sin(angle);
+  // prettier-ignore
+  this.mult([1, 0, 0, 0, 
```

Comment:
I'm not sure I understand why these calculations are done here instead of deferring to `rotate(angle, [0, 0, 1])`? and likewise for rotateY, rotateZ.